proto/src/rr/domain/label.rs

```diff
@@ -0,0 +1,307 @@
+#[derive(Clone, Eq)]
+pub struct Label(Rc<[u8]>);
+
+impl Label {
+    /// These must only be ASCII, with unicode encoded to PunyCode
+    ///
+    /// This uses the bytes as raw ascii values.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, ProtoError> {
+        Ok(Label(Rc::from(bytes)))
+    }
+
+    /// Converts this label to lowercase
+    pub fn to_lowercase(&self) -> Self {
+        // TODO: replace case conversion when (ascii_ctype #39658) stabilizes
+        if let Some((idx, _)) = self.0.iter().enumerate().find(|&(_, c)| *c != c.to_ascii_lowercase()) {
+            let mut lower_label: Vec<u8> = self.0.to_vec();
+            lower_label[idx..].make_ascii_lowercase();
+            Label(Rc::from(lower_label))
+        } else {
+            self.clone()
+        }
+    }
+
+    /// Returns true if this label is the wildcard, '*', label
+    pub fn is_wildcard(&self) -> bool {
+        self.as_bytes() == WILDCARD
+    }
+
+    /// Returns the lenght in bytes of this label
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns the raw bytes of the label, this is good for writing to the wire.
+    ///
+    /// See [`Display`] for presentation version (unescaped from punycode, etc)
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    /// Performs the equivelence operation disregarding case
+    pub fn eq_ignore_ascii_case(&self, other: &Self) -> bool {
+        self.0.eq_ignore_ascii_case(&other.0)
+    }
+
+    /// compares with the other label, ignoring case
+    pub fn cmp_with_f<F: LabelCmp>(&self, other: &Self) -> Ordering {
+        let s = self.0.iter();
+        let o = other.0.iter();
+        
+        for (s, o) in s.zip(o) {
+            match F::cmp_u8(*s, *o) {
+                Ordering::Equal => continue,
+                not_eq => return not_eq,
+            }
+        }
+
+        self.0.len().cmp(&other.0.len())
+    }
+}
+
+impl AsRef<[u8]> for Label {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl Borrow<[u8]> for Label {
+    fn borrow(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl FromStr for Label {
+    type Err = ProtoError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.as_bytes() == WILDCARD {
+            return Ok(Label(Rc::from(WILDCARD.to_vec())))
+        }
+
+        // if it's all ascii, don't perform the unicode conversion
+        if s.is_ascii() {
+            return Label::from_bytes(s.as_bytes())
+        }
+
+        match uts46::to_ascii(s, uts46::Flags{
+            use_std3_ascii_rules: true, 
+            transitional_processing: true, 
+            verify_dns_length: true}) {
+            Ok(puny) => {
+                Ok(Label(Rc::from(puny.as_bytes())))
+            }
+            Err(e) => {
+                debug!("Label contains invalid characters, treating as binary: {:?}", e);
+                Ok(Label(Rc::from(s.as_bytes())))
```

Comment:
I'm considering removing this fall back and returning an error. If a binary name is desired then it can be forcefully introduced with the `from_bytes` method.